Mathlib/Combinatorics/Optimization/ValuedCSP.lean

```diff
@@ -0,0 +1,216 @@
+/-
+Copyright (c) 2023 Martin Dvorak. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Martin Dvorak
+-/
+import Mathlib.Algebra.Order.Monoid.Defs
+import Mathlib.Data.Finset.Basic
+import Mathlib.Data.Fin.VecNotation
+import Mathlib.Data.Rat.Order -- will be trimmed after deleting examples
+import Mathlib.Tactic.Linarith
+import Mathlib.Tactic.Cases
+import Mathlib.Data.Bool.Basic
+
+/-!
+
+# General-Valued Constraint Satisfaction Problems
+
+General-Valued CSP is a very broad class of problems in discrete optimization.
+General-Valued CSP subsumes Min-Cost-Hom (including 3-SAT for example) and Finite-Valued CSP.
+
+## Main definitions
+* `ValuedCspTemplate`: A VCSP template; fixes a domain, a codomain, and allowed cost functions.
+* `ValuedCspTerm`: One summand in a VCSP instance; calls a concrete function from given template.
+* `ValuedCspTerm.evalSolution`: An evaluation of the VCSP term for given solution.
+* `ValuedCspInstance`: An instance of a VCSP problem over given template.
+* `ValuedCspInstance.evalSolution`: An evaluation of the VCSP instance for given solution.
+* `ValuedCspInstance.optimumSolution`: Is given solution a minimum of the VCSP instance?
+
+## References
+* [D. A. Cohen, M. C. Cooper, P. Creed, P. G. Jeavons, S. Živný,
+   *An Algebraic Theory of Complexity for Discrete Optimisation*][cohen2012]
+
+-/
+
+def n1ary_of_unary {α β : Type _} (f : α → β) : (Fin 1 → α) → β :=
```

Comment:
A general comment; where possible, `Type _` should be `Type*` throughout.

Reply:
Will be `Type` instead.

Reply:
Or is it something that the comment by @fpvandoorn does not apply to?

Reply:
Actually `n1aryOfUnary` and  `n2aryOfBinary` will be with `Type*`. Can all definitions about VCSP be in `Type` tho?

Reply:
No, you should almost never use `Type` without a `*`. I fully agree with Eric here, you should use `Type*` by default.

The reason that `D : Type` in the structure is sort-of acceptable, is 
* It's a structure field, and adding a new universe there makes the structure depend on another universe, which is annoying
* It's almost always a finite indexing type, so should live in the bottom universe. It's still possible that using `Type` will come to bite us later.
But in short: use `Type*` everywhere else.

Reply:
Ah, I get it. I will refector everything but `D` to `Type*` as @eric-wieser suggested.